File: android/app/src/main/java/org/onemindlabs/airhop/transport/Framing.kt
package org.onemindlabs.airhop.transport

// The stream framing every TCP link uses, Wi-Fi Aware and LAN alike: a 4-byte
// big-endian length, then the bytes. An empty frame is the heartbeat. Pure, so
// the read loops and the tests share one definition of a valid prefix.
internal object Framing {
    // One 64 KiB file chunk plus the length prefix.
    const val MAX_FRAME = 65544
    const val PREFIX_BYTES = 4

    fun encode(data: ByteArray): ByteArray {
        val frame = ByteArray(PREFIX_BYTES + data.size)
        val len = data.size
        frame[0] = (len shr 24).toByte()
        frame[1] = (len shr 16).toByte()
        frame[2] = (len shr 8).toByte()
        frame[3] = len.toByte()
        data.copyInto(frame, PREFIX_BYTES)
        return frame
    }

    // The body length a prefix announces, or null when the prefix cannot be
    // trusted: a high bit set (negative as an Int) or a claim past MAX_FRAME,
    // either of which means the stream is not ours or has lost sync.
    fun length(prefix: ByteArray): Int? {
        if (prefix.size < PREFIX_BYTES) return null
        val len =
            ((prefix[0].toInt() and 0xff) shl 24) or
                ((prefix[1].toInt() and 0xff) shl 16) or
                ((prefix[2].toInt() and 0xff) shl 8) or
                (prefix[3].toInt() and 0xff)
        if (len < 0 || len > MAX_FRAME) return null
        return len
    }
}

File: ios/Airhop/Transport/Framing.swift
// The stream framing every TCP link uses, Wi-Fi Aware and LAN alike: a 4-byte
// big-endian length, then the bytes. An empty frame is the heartbeat. Pure, so
// the read loops and the tests share one definition of a valid prefix.
//
// Big-endian by hand rather than reading a `UInt32`, which comes back in host
// order and is byte-swapped on every device this runs on.
import Foundation

enum Framing {
    // One 64 KiB file chunk plus the length prefix.
    static let maxFrame = 65_544
    static let prefixBytes = 4

    static func encode(_ payload: Data) -> Data {
        let length = UInt32(payload.count)
        var out = Data(capacity: prefixBytes + payload.count)
        out.append(UInt8((length >> 24) & 0xff))
        out.append(UInt8((length >> 16) & 0xff))
        out.append(UInt8((length >> 8) & 0xff))
        out.append(UInt8(length & 0xff))
        out.append(payload)
        return out
    }

    // The body length a prefix announces, or nil when the prefix cannot be
    // trusted: a claim past maxFrame means the stream is not ours or has lost
    // sync.
    static func length(_ prefix: Data) -> Int? {
        guard prefix.count == prefixBytes else { return nil }
        let bytes = [UInt8](prefix)
        let length =
            (Int(bytes[0]) << 24) | (Int(bytes[1]) << 16) | (Int(bytes[2]) << 8) | Int(bytes[3])
        guard length <= maxFrame else { return nil }
        return length
    }
}
